control-plane/agents/core/src/core/specs.rs

```diff
@@ -35,6 +37,416 @@ enum SpecError {
+        registry: &Registry,
+    ) -> Result<R, SvcError>
+    where
+        Self: SpecTransaction<O>,
+        Self: StorableObject,
+    {
+        let key = locked_spec.lock().await.key();
+        match result {
+            Ok(val) => {
+                let mut spec_clone = locked_spec.lock().await.clone();
+                spec_clone.commit_op();
+                let deleted = registry.delete_kv(&key.key()).await;
+                match deleted {
+                    Ok(_) => {
+                        Self::remove_spec(locked_spec, registry).await;
+                        let mut spec = locked_spec.lock().await;
+                        spec.commit_op();
+                        Ok(val)
+                    }
+                    Err(error) => {
+                        let mut spec = locked_spec.lock().await;
+                        spec.set_op_result(true);
+                        Err(error)
+                    }
+                }
+            }
+            Err(error) => {
+                let mut spec_clone = locked_spec.lock().await.clone();
+                spec_clone.clear_op();
+                let stored = registry.store_obj(&spec_clone).await;
+                let mut spec = locked_spec.lock().await;
+                match stored {
+                    Ok(_) => {
+                        spec.clear_op();
+                        Err(error)
+                    }
+                    Err(error) => {
+                        spec.set_op_result(false);
+                        Err(error)
+                    }
+                }
+            }
+        }
+    }
+
+    /// Start a destroy operation and attempt to log the transaction to the store.
+    /// In case of error, the log is undone and an error is returned.
+    async fn start_update(
+        registry: &Registry,
+        locked_spec: &Arc<Mutex<Self>>,
+        status: &Self::Status,
+        update_operation: Self::UpdateOp,
+    ) -> Result<Self, SvcError>
+    where
+        Self: PartialEq<Self::Status>,
+        Self: SpecTransaction<Self::UpdateOp>,
+        Self: StorableObject,
+    {
+        let mut spec = locked_spec.lock().await;
+        let spec_clone = spec.start_update_inner(status, update_operation, false)?;
+        drop(spec);
+
+        if let Err(error) = registry.store_obj(&spec_clone).await {
+            let mut spec = locked_spec.lock().await;
+            spec.clear_op();
+            Err(error)
+        } else {
+            Ok(spec_clone)
+        }
+    }
+
+    /// Checks that the object ready to accept a new update operation
+    fn start_update_inner(
+        &mut self,
+        status: &Self::Status,
+        operation: Self::UpdateOp,
+        reconciling: bool,
+    ) -> Result<Self, SvcError>
+    where
+        Self: PartialEq<Self::Status>,
+    {
+        // we're busy right now, try again later
+        let _ = self.busy()?;
+
+        if !self.state().created() {
+            Err(SvcError::PendingCreation {
```

Comment:
If it's not created can we guarantee that creation is pending? Is it possible that it is not created and nothing is trying to create it?

Reply:
the reconciler should be the one trying to do something with it, unless we want to define a state where we say we've given up on the resource and we'll stop trying to reconcile and the resource will never become usable again?

Reply:
Done, split up the creating/deleting